src/impl.cxx

```diff
@@ -733,6 +734,26 @@ namespace ipr {
             return compare(lhs, rhs.node);
          }
 
+         template<class T,
+            class = std::enable_if_t<!std::is_same_v<
+              std::remove_reference_t<typename Unary<T>::Arg_type>,
+              const ipr::Sequence<ipr::Type>>>>
+         int operator()(const Unary<T>& lhs,
```

Comment:
When we get to apply SFINAE, that means we are doing something quite not right.  Could you elaborate on why we would need to do this?  Do you see alternatives?

Reply:
Interestingly, MSVC and Clang compiled the code without any trouble. But according to GCC, this call is ambiguous with the `ipr::Sequence<ipr::Type>` overloads. I added the SFINAE as a workaround to the GCC compilation error. I did not look too deeply into it yet, but I suspect that this might be a bug in GCC.

Reply:
Thanks!  Sometimes, SFINAE is a necessary evil but I tend to recommend to avoid it.  If we were using C++20, we could be using concepts and the constraints would be most apparent.  However, we should be able to write this library simply and elegantly even in C++03.  So, let's me ponder on your the issue a a bit more, thanks to your explanation.

Reply:
I see that IPR is now compiled in C++20 mode. Did you have a chance to look into it if there is a nicer way to select the right overload?

Reply:
> (On GCC as other compilers seem to pick the right one)

So, this is a recent GCC bug?  Which version of GCC?

The more I look at the patch, the more I become uncomfortable with it.  The PR claims that it is to address lack of uniqueness of `Linkage`.  Yet, it goes on adding SFINAE on `Sequence<ipr::Type>`.  That does not logically sound right, even if it makes the GCC compiler comply.

Reply:
> The PR claims that it is to address lack of uniqueness of Linkage. Yet, it goes on adding SFINAE on Sequence<ipr::Type>. 

No. The SFINAE is not the fix. The fix is adding an overload. Unfortunately, GCC thinks this is ambiguous. The SFINAE is working that one around. 

I did not check multiple GCC versions, the one that fails is the one in the CI. I will check its version once I'm back to work (tomorrow).

Reply:
> The fix is adding an overload. 

In the diff, I see only two additional overloads: they are symmetric of each other and are constraining `Sequence<ipr::Type>`.  Am I reading the diff wrong?  It GH playing trick on me?

Reply:
Yeah, that is right, but I still have problem understanding your point here.

> The PR claims that it is to address lack of uniqueness of Linkage.

If you do not trust that the PR does what it claims, you can run the attached test without the two added overload and see what happens, you can debug into the code and see why. The bug happens on all compiler versions. If I ad the overloads, the bug is fixed on Clang and MSVC but the compilation fails on GCC. But this is encoded in the commit history of the PR. The first commit adds the overload, the second one adds the SFINAE and the commit messages describe why did that happen. 

In the meantime, I looked at the error. In gcc 9.3, without the SFINAE I get the following error message after adding the new overloads:

```
[08:23:54] xazax@LAPTOP-HF5CMH47 /home/xazax/ipr/build
> make
Scanning dependencies of target ipr
[ 11%] Building CXX object CMakeFiles/ipr.dir/src/impl.cxx.o
In file included from /home/xazax/ipr/include/ipr/interface:16,
                 from /home/xazax/ipr/include/ipr/impl:19,
                 from /home/xazax/ipr/src/impl.cxx:7:
/home/xazax/ipr/include/ipr/utility: In instantiation of ‘T* ipr::util::rb_tree::container<T>::insert(const Key&, Comp) [with Key = ipr::Sequence<ipr::Type>; Comp = ipr::impl::unary_compare; T = ipr::impl::Unary<ipr::impl::Type<ipr::Product> >]’:
/home/xazax/ipr/src/impl.cxx:935:53:   required from here
/home/xazax/ipr/include/ipr/utility:344:35: error: call of ‘(ipr::impl::unary_compare) (const ipr::Sequence<ipr::Type>&, ipr::impl::Unary<ipr::impl::Type<ipr::Product> >&)’ is ambiguous
  344 |                int ordering = comp(key, where->data);
      |                               ~~~~^~~~~~~~~~~~~~~~~~
/home/xazax/ipr/src/impl.cxx:745:14: note: candidate: ‘int ipr::impl::unary_compare::operator()(const typename ipr::impl::Unary<T>::Arg_type&, const ipr::impl::Unary<T>&) const [with T = ipr::impl::Type<ipr::Product>; typename ipr::impl::Unary<T>::Arg_type = const ipr::Sequence<ipr::Type>&]’
  745 |          int operator()(const typename Unary<T>::Arg_type& lhs,
      |              ^~~~~~~~
/home/xazax/ipr/src/impl.cxx:761:14: note: candidate: ‘int ipr::impl::unary_compare::operator()(const ipr::Sequence<ipr::Type>&, const ipr::impl::Unary<T>&) const [with T = ipr::impl::Type<ipr::Product>]’
  761 |          int operator()(const ipr::Sequence<ipr::Type>& lhs,
      |              ^~~~~~~~
In file included from /home/xazax/ipr/include/ipr/interface:16,
                 from /home/xazax/ipr/include/ipr/impl:19,
                 from /home/xazax/ipr/src/impl.cxx:7:
/home/xazax/ipr/include/ipr/utility: In instantiation of ‘T* ipr::util::rb_tree::container<T>::insert(const Key&, Comp) [with Key = ipr::Sequence<ipr::Type>; Comp = ipr::impl::unary_compare; T = ipr::impl::Unary<ipr::impl::Type<ipr::Sum> >
’:
/home/xazax/ipr/src/impl.cxx:957:49:   required from here
/home/xazax/ipr/include/ipr/utility:344:35: error: call of ‘(ipr::impl::unary_compare) (const ipr::Sequence<ipr::Type>&, ipr::impl::Unary<ipr::impl::Type<ipr::Sum> >&)’ is ambiguous
  344 |                int ordering = comp(key, where->data);
      |                               ~~~~^~~~~~~~~~~~~~~~~~
/home/xazax/ipr/src/impl.cxx:745:14: note: candidate: ‘int ipr::impl::unary_compare::operator()(const typename ipr::impl::Unary<T>::Arg_type&, const ipr::impl::Unary<T>&) const [with T = ipr::impl::Type<ipr::Sum>; typename ipr::impl::Unary<T>::Arg_type = const ipr::Sequence<ipr::Type>&]’
  745 |          int operator()(const typename Unary<T>::Arg_type& lhs,
      |              ^~~~~~~~
/home/xazax/ipr/src/impl.cxx:761:14: note: candidate: ‘int ipr::impl::unary_compare::operator()(const ipr::Sequence<ipr::Type>&, const ipr::impl::Unary<T>&) const [with T = ipr::impl::Type<ipr::Sum>]’
  761 |          int operator()(const ipr::Sequence<ipr::Type>& lhs,
      |              ^~~~~~~~
make[2]: *** [CMakeFiles/ipr.dir/build.make:76: CMakeFiles/ipr.dir/src/impl.cxx.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:114: CMakeFiles/ipr.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

Reply:
> If you do not trust that the PR does what it claims, you can run the attached test without the two added overload and see what happens, you can debug into the code and see why. The bug happens on all compiler versions. If I ad the overloads, the bug is fixed on Clang and MSVC but the compilation fails on GCC. But this is encoded in the commit history of the PR. The first commit adds the overload, the second one adds the SFINAE and the commit messages describe why did that happen.

The issue here isn't that of a trust or not.  The issue is the need for the SFINAE.  Furthermore, the fact that there is a disagreement between the three compilers on addition of overload without constraint calls for understanding further what is going on.  In addition to the SFINAE being a red flag,

Reply:
I completely agree. Unfortunately, I am very uncomfortable working with large overload sets, so I have no idea how to make this approach feel more right. I did not want to do larger changes (which would make the review process potentially harder), but I think if it was my project, I would try to split `unary_compare` up into other, smaller classes to avoid big/complicated overload sets. Here, the source of the error for node duplication was that we ended up picking a too generic overload. It looks like with big overload sets it is relatively easy to write code that compiles but does not do what one would expect.

Reply:
I will give that a try tomorrow.

Reply:
> I will give that a try tomorrow.

Thanks!

In general, we should prioritize the right thing to do over the size of the diff.  Of course, sometimes, the right thing to do isn't always obvious...

Reply:
Splitting up the overload set seems to solve the problem. Updated the PR.

Reply:
> Splitting up the overload set seems to solve the problem. Updated the PR.

Thanks!